### ISSUE TITLE
Wall-clock profiling

### ISSUE THREAD

**Issue by ksafonov** [jvm-profiling-tools/async-profiler#69]
It may be helpful to measure not only the time that method spent consuming the CPU but also the time method was waiting for lock or I/O. Please implement 'wall clock' profiling mode.

**Comment by nitsanw** [comment]
What you describe is the sort of thing implemented by instrumenting/tracing profilers. At this time async-profiler is a sampling profiler and AFAIK there's no intention to expand it's capabilities in the direction you describe (instrucmenting timestamps before/after method execution to provide time spent in each method).
Having said that async-profiler does lock profiling, and allocation profiling which both use tracepoints to instrument particular JVM Runtime events. There's certainly interest in tracing off-cpu time which can be done by tracing scheduler events and might be worth looking into.

**Comment by apangin** [comment]
Actually @ksafonov asked me about this feature and I've approved submitting this request, since there are definitely use cases that can benefit from wall-clock profiling. I'm also interested in this, for example, for profiling application start-up time.

I believe sampling profiler can solve this task to some extent. Basically, good old `GetAllStackTraces` can do the thing, though with a safepoint bias. I'm investigating how the similar functionality can be achieved without safepoint bias.

**Comment by nitsanw** [comment]
@apangin generically computing time in method for all methods is quite an invasive change to the program under test. I assume this is not on the cards here? maybe I just misunderstood what @ksafonov is after?

**Comment by apangin** [comment]
Certainly I'm not going to compute time spent in all methods in async-profiler. However, sampling techniques can be useful for wall-clock profiling, too. We just need to sample idle threads similarly to busy threads. That's what classic sampling profilers do.

I'm not sure if this matches original @ksafonov's intention, but this can definitely solve some scope of profiling problems.

**Comment by nitsanw** [comment]
If we want off CPU timing I think all the data we need should be in the data set:
 1. Get stacks for all blocked threads
 2. Trace schedule events throughout the profile period (can use current method of walking stack)
This is a tracing style rather than sampling approach and I think is quite valuable.

**Comment by apangin** [comment]
I've added `-e wall` option in [wall-clock](https://github.com/jvm-profiling-tools/async-profiler/compare/wall-clock) branch. I'll be glad to get early feedback before the feature goes to main line.

**Comment by retronym** [comment]
There is a build failure on MacOS

```
$ make clean all
rm -rf build
mkdir -p build
g++ -O2 -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -DPROFILER_VERSION=\"1.5-ea\" -I/Users/jz/.jabba/jdk/1.8.172/Contents/Home/include -I/Users/jz/.jabba/jdk/1.8.172/Contents/Home/include/darwin -fPIC -shared -o build/libasyncProfiler.so src/*.cpp -ldl -lpthread
In file included from src/profiler.cpp:28:
src/wallClock.h:38:5: warning: control reaches end of non-void function [-Wreturn-type]
    }
    ^
1 warning generated.
src/wallClock.cpp:22:10: fatal error: 'sys/eventfd.h' file not found
#include <sys/eventfd.h>
         ^~~~~~~~~~~~~~~
1 error generated.
make: *** [build/libasyncProfiler.so] Error 1

$ uname -a
Darwin iMac2016.local 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
```

**Comment by apangin** [comment]
Yes, it works on Linux only right now. I'm investigating how to implement the similar functionality on macOS.

**Comment by calvin1978** [comment]
any update of this issue ? 

we really want to profile which method costed most time, just like restful or db call.

**Comment by apangin** [comment]
@calvin1978 Have you tried a version from [wall-clock](https://github.com/jvm-profiling-tools/async-profiler/tree/wall-clock) branch? The feature is ready both on Linux and macOS. The feedback will help to integrate or improve the feature sooner.

**Comment by calvin1978** [comment]
> @calvin1978 Have you tried a version from [wall-clock](https://github.com/jvm-profiling-tools/async-profiler/tree/wall-clock) branch? The feature is ready both on Linux and macOS. The feedback will help to integrate or improve the feature sooner.

ok, i will test it next week.

**Comment by calvin1978** [comment]
@apangin shall we skip some famous method which just wait for incoming request, like 

java/util/concurrent/ThreadPoolExecutor.getTask
sun/nio/ch/EPollArrayWrapper.epollWait

i remember that some other profilers  skip them

**Comment by calvin1978** [comment]
@apangin  it works well in linux, and can you add one more options to skip some famous methods which just waiting for incoming request, like

java/util/concurrent/ThreadPoolExecutor.getTask
sun/nio/ch/EPollArrayWrapper.epollWait

these method occupy too much space , and i remember that some other profiler will skip them.

**Comment by calvin1978** [comment]
./profiler.sh list 102788
Perf events:
  cpu
  page-faults
  context-switches
  cycles
  instructions
  cache-references
  cache-misses
  branches
  branch-misses
  bus-cycles
  L1-dcache-load-misses
  LLC-load-misses
  dTLB-load-misses
  mem:breakpoint
  trace:tracepoint
Java events:
  alloc
  lock

wall didn't be listed in Java events.

**Comment by apangin** [comment]
@calvin1978 Thank you for the feedback. I'll take your comments into account.

**Comment by calvin1978** [comment]
let's skip the further request like  omit "ThreadPoolExecutor.getTask", can you merge the branch back to master?

 so we can build our snapshot with wall clock  feature  and also bug fix for clock profiling with JDK7

**Comment by apangin** [comment]
Thank you @retronym. The problem is understood. I could probably add a switch to exclude VM threads. Thread pools are not that simple though. They are not always idle, and it would be unfair to include only their running time - or this wouldn't be "wall-clock" mode anymore.
It's generally a good idea to run "wall-clock" profiling with `-t` option so that irrelevant threads could be easily grepped out afterwards.

**Comment by calvin1978** [comment]
> @calvin1978 There are some architectural conflicts that need to be resolved before merging the branch into master. But I just did the reverse: synced wall-clock branch with current master.

thanks @apangin , so i can build a snapshot include all bug fixing use this branch , right ?

**Comment by apangin** [comment]
@calvin1978 Yes, you may proceed with this branch - now it contains all bugfixes from the upstream.

**Comment by apangin** [comment]
@oehme The problem is there are no "idle" threads in off-cpu profiling. Instead there are threads you are interested in and threads you are not interested in. Nobody knows how to distinguish one from another except yourself :)

For example, consider a thread waiting on a `Condition` object. This can be either an idle thread in some large ThreadPool, or a worker thread that can't proceed until a dependent task finishes. The stack traces will be the same, but in the former case this thread is irrelevant, while in the latter case this thread blocks execution of business logic. Similarly, a thread blocked on `Socket.read` may be completely irrelevant (e.g. when listening solely for a shutdown message) or this may be a thread waiting for an important response.

I believe it should be a user decision to choose which events or threads they are interested in. My intention was to provide some way to filter relevant threads or samples. I want to collect real use cases and suggestions: how would you like this filtering feature to be implemented in order to be useful in particular scenarios?

**Comment by oehme** [comment]
Right, looking at the profile it "looks obvious" that I wouldn't care about the GC thread waiting for stuff to do, but it's just waiting on a condition.

For my purposes it would probably be enough to be able to filter threads with a naming pattern or a thread group.

**Comment by ksafonov** [comment]
Just a stupid question: is the full-GC time included into wall time? I.e. if my thread is blocked on a lock for 5 ms, then there comes a 200-ms full GC pause, then my thread is sleeping for another 5ms, will I get 210 ms or 10 ms as the result?

**Comment by oehme** [comment]
All threads are profiled, that includes GC. However, profiling is not based on time, but samples.

**Comment by apangin** [comment]
@ksafonov It's an interesting question actually! It made me think of the aspects I haven't realized before.
1. Wall-clock profiler is still a sampling profiler. It does not show time spent inside a method, but the time can be *estimated* basing on the number of samples.
2. Sampler itself is not affected by GC, i.e. it produces signals in equal periods of time, no matter whether JVM runs garbage collection or not.
3. GC threads are sampled equally along with application threads.
4. The problem is **JVM isn't ready** to walk through Java stacks while GC is active. `AsyncGetCallTrace` API does not support that and returns an error in such cases.
5. Meanwhile profiler can still walk through native stacks, so you'll see a partial stack trace containing `SafepointSynchronize::block(JavaThread*)` or similar.
6. At the same time application threads are known to be at safepoint while GC is running, so it should be possible to obtain correct Java stack by other means. I've created #202 to address this issue.

**Comment by apangin** [comment]
UPDATE: Now Java stacks are displayed correctly during wall-clock profiling even when GC is active.
